### ISSUE TITLE
An equivalent of Apache's +ExportCertData in Caddy

### ISSUE THREAD

**Issue by Marian-Kechlibar** [caddyserver/caddy#6713]
Apache Webserver's mod_ssl has an optional configuration flag +ExportCertData, which is turned off by default (for performance reasons), but when turned on, it populates the superglobals with variables containing the current server and client certificate. To cite from the docs:

_When this option is enabled, additional CGI/SSI environment variables are created: SSL_SERVER_CERT, SSL_CLIENT_CERT and SSL_CLIENT_CERT_CHAIN_n (with n = 0,1,2,..). These contain the PEM-encoded X.509 Certificates of server and client for the current HTTPS connection and can be used by CGI scripts for deeper Certificate checking. Additionally all other certificates of the client certificate chain are provided, too. This bloats up the environment a little bit which is why you have to use this option to enable it on demand._

This would be useful in Caddy too, as it would make migration of Apache-optimized apps to Caddy more straightforward. We had a debate with @francislavoie about the functionality here and I made case that it makes sense to have access to the server-side cert as well.

https://caddy.community/t/populating-server-server-cert-and-server-client-cert/26521

He recommended me to start an Issue on GitHub, so here I am. I know next to nothing about Caddy code, so I dare not submit a Pull request for that functionality.

**Comment by mholt** [comment]
Thanks -- to clarify, it looks like, based on the forum topic, the request is actually for placeholders containing server cert info.

**Comment by Marian-Kechlibar** [comment]
Yes. it is. 

Meanwhile, I was able to use the placeholder for the client certificate, but with some unexpected behavior, which may warrant its own issue...

Namely, when I enter the following into my Caddyfile:

	request_header +X-Tls-Client-Cert-Der-Base64 {http.request.tls.client.certificate_der_base64}

the request gets added even if there is no client cert provided, and populated with the literal  "{http.request.tls.client.certificate_der_base64}"

That is unexpected for me. I would expect the header either to be absent, or empty in that case.  (If HTTP allows empty headers; IDK). Should I start another topic over that purpose.

**Comment by mholt** [comment]
Ah, that's because in the header directive(s) only _known_ placeholders are replaced. If there is no client certificate, the placeholder is not set.

We can change this in one of two ways:

- Always replace placeholder-looking substrings, but this is problematic (syntactically) for values that aren't intended to be placeholders (e.g. JSON)
- Replacers return two values: the value, and whether the key (placeholder name) is known. We are evidently returning `false` to known, when there is no client cert. It is kind of debatable whether that is correct. We could return `true` as in, we know this placeholder, but do we really if there is no client auth?

I'm leaning towards the second though, if we do make a change.

**Comment by steffenbusch** [comment]
Can you try this:

```Caddyfile

@hasClientCert {
       not vars {http.request.tls.client.certificate_der_base64} ""
}
request_header @hasClientCert +X-Tls-Client-Cert-Der-Base64 {http.request.tls.client.certificate_der_base64}
```

**Comment by mholt** [comment]
Thanks for the workaround @steffenbusch !

I suppose we could still change the placeholders behavior in the case of client auth missing.

**Comment by Ko496-glitch** [comment]
@mholt is this issue still open? because i would like to take a try on this.

**Comment by mholt** [comment]
I would start in here:

https://github.com/caddyserver/caddy/blob/1bd567d7ad41d5509e2aa60cf36e749f195ad83c/modules/caddyhttp/replacer.go#L399-L492

**Comment by Ko496-glitch** [comment]
@mholt So i did some research and found that apache mod_ssl env variables list is huge so do you want me to work on most common use env variables or each of those that mod_ssl provides in apache. I am pasting the screenshot for reference. 
Thanks

<img width="1134" alt="Image" src="https://github.com/user-attachments/assets/d37c3583-d1cd-4550-b8a2-96b25c3add90" />

**Comment by mohammed90** [comment]
I think it's best to stay conservative in what we add, and only add what's needed.

**Comment by mholt** [comment]
Yeah, already have some of those; I would stick to just the few that this issue is requesting, but even before that, if you want, we can make this adjustment: https://github.com/caddyserver/caddy/issues/6713#issuecomment-2518464583

**Comment by Ko496-glitch** [comment]
I agree that the second approach makes sense. By still parsing the placeholder and returning a value like "No Client cert provided" when no .ClientCert is available during runtime, we can gracefully handle scenarios where the user is not doing mTLS.

**Comment by Ko496-glitch** [comment]
@mholt I’ve implemented functionality to handle both client and server certificate placeholders and environment variables.
From my understanding the code has following implementations -:
1) Adds +ExportCertData functionality to populate placeholders or a return map for CGI or other integrations.
2) Ensures placeholders are populated with default values when no client certificate is provided (e.g., "No Client Certificate Provided").
3) Handles server and client certificate placeholders using Caddy's replacer.
I am confuse about how to post the code so looking for your support on this once reviewed properly.

**Comment by Marian-Kechlibar** [comment]
Just a small comment, if it is desirable to be compatible with Apache, SSL_CLIENT_CERT should default to null if ńot present in the handshake, and no other placeholder default is given in the Caddyfile.

**Comment by Ko496-glitch** [comment]
@Marian-Kechlibar  just to be clear do you mean that when we only doing TLS(disable mTLS) SSL_CLIENT_CERT placeholder will return something like this 

func handleMTLSEnabledWithExport( 
    req *http.Request,
    connState *tls.ConnectionState, 
    exportCertData bool,
) map[string]string{
 certDetails := make(map[string]string)

// Get the Caddy replacer from the request context
repl, _ := req.Context().Value(caddyhttp.ReplacerCtxKey).(*caddyhttp.Replacer)

// Helper function to set placeholders or map values
setData := func(key, value string) {
    if exportCertData && repl != nil {
        repl.Set(key, value)
    } else {
        certDetails[key] = value
    }
}

// This part handles Server certificate details
if connState != nil && len(connState.PeerCertificates) > 0 {
    serverCert := connState.PeerCertificates[0]
    setData("SSL_SERVER_CERT", string(pem.EncodeToMemory(&pem.Block{
        Type:  "CERTIFICATE",
        Bytes: serverCert.Raw,
    })))
    setData("SSL_SERVER_SUBJECT", serverCert.Subject.String())
    setData("SSL_SERVER_ISSUER", serverCert.Issuer.String())
} else {
    setData("SSL_SERVER_CERT", "No Server Certificate Found")
    setData("SSL_SERVER_SUBJECT", "No Server Certificate Subject")
    setData("SSL_SERVER_ISSUER", "No Server Certificate Issuer")
}

// This part handles the Client certificate details
if connState != nil && len(connState.VerifiedChains) > 0 {
    clientCert := connState.VerifiedChains[0][0]
    setData("SSL_CLIENT_CERT", string(pem.EncodeToMemory(&pem.Block{
        Type:  "CERTIFICATE",
        Bytes: clientCert.Raw,
    })))
    setData("SSL_CLIENT_SUBJECT", clientCert.Subject.String())
    setData("SSL_CLIENT_ISSUER", clientCert.Issuer.String())
} else {
    setData("SSL_CLIENT_CERT", "No Client Certificate Provided")
    setData("SSL_CLIENT_SUBJECT", "No Client Certificate Subject")
    setData("SSL_CLIENT_ISSUER", "No Client Certificate Issuer")
}

return certDetails

**Comment by Ko496-glitch** [comment]
i will open a PR  with changes in replacer.go we can always review and add or subtract during review.
